src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Extensions/BaseTypeIntermediateNode.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions;
+
+public sealed class BaseTypeIntermediateNode : ExtensionIntermediateNode
+{
+    const string ModelGenericParameter = "<TModel>";
+
+    public override IntermediateNodeCollection Children { get; } = IntermediateNodeCollection.ReadOnly;
+
+    public BaseTypeIntermediateNode(string baseType, SourceSpan? location)
+    {
+        if (baseType.EndsWith(ModelGenericParameter, System.StringComparison.Ordinal))
+        {
+            BaseType = IntermediateToken.CreateCSharpToken(baseType[0..^ModelGenericParameter.Length]);
+            GreaterThan = IntermediateToken.CreateCSharpToken("<");
+            ModelType = IntermediateToken.CreateCSharpToken("TModel");
+            LessThan = IntermediateToken.CreateCSharpToken(">");
+
+            if (location.HasValue)
+            {
+                var openBracketPosition = baseType.Length - ModelGenericParameter.Length;
+                BaseType.Source = location.Value.Slice(0, openBracketPosition);
+                GreaterThan.Source = location.Value.Slice(openBracketPosition, openBracketPosition + 1);
+                ModelType.Source = location.Value.Slice(openBracketPosition + 1, baseType.Length - 1);
+                LessThan.Source = location.Value.Slice(baseType.Length - 1, baseType.Length);
+            }
+        }
+        else
+        {
+            BaseType = IntermediateToken.CreateCSharpToken(baseType, location);  
+        }
+        Source = location;
+    }
+
+    public IntermediateToken BaseType { get; set; }
+
+    public IntermediateToken? GreaterThan { get; set; }
```

Comment:
So if these can be null, can we simply use BaseTypeIntermediateNode everywhere like:
- in MVC with `<TModel>`, use BaseTypeIntermediateNode with BaseType, GreaterThan, ModelType, LessThan set;
- otherwise, use BaseTypeIntermediateNode with only BaseType set and the others being `null`.

And that might hopefully simplify the handling where we now have to conditionally handle BaseTypeIntermediateNode vs IntermediateToken.

I see you suggested something similar:

> I don't love the fact that we have to type check the base type now. It's probably possible to have a base 'BaseType' node that MVC then extends, so we get a single type, but its used in a whole bunch of places and it didn't seem worth it. 

But I was wondering if we could do that without the inheritance.

Also

> but its used in a whole bunch of places and it didn't seem worth it

does that mean we would need to refactor a lot of places if we used BaseTypeIntermediateNode for ClassDeclarationIntermediateNode.BaseType? I guess than I'm fine with not doing that. But aren't we touching all those places now anyway since we went from IntermediateToken to IntermediateNode?

Reply:
Yeah, I ended up going that route of just having it be a type for everything. I was worried it would be weird for the things that don't use model, but it actually works quite well I think, and is way simpler.